anesthetic/samples.py

```diff
@@ -209,21 +242,55 @@ def plot_2d(self, axes, *args, **kwargs):
             Pandas array of axes objects
 
         """
-        types = kwargs.pop('types', ['kde', 'scatter'])
+        types = kwargs.pop('types', {'diagonal': 'kde', 'lower': 'kde'})
+        diagonal = kwargs.pop('diagonal', True)
+        if isinstance(types, list) or isinstance(types, str):
+            from warnings import warn
+            warn("MCMCSamples.plot_2d's argument 'types' might stop accepting "
+                 "str or list(str) as input in the future. It takes a "
+                 "dictionary as input, now, with keys 'diagonal' for the 1D "
+                 "plots and 'lower' and 'upper' for the 2D plots. 'diagonal' "
+                 "accepts the values 'kde' or 'hist' and both 'lower' and "
+                 "'upper' accept the values 'kde' or 'scatter'. "
+                 "Default: {'diagonal': 'kde', 'lower': 'kde'}.",
+                 FutureWarning)
+            if isinstance(types, str) and diagonal:
+                types = {'diagonal': types, 'lower': types}
+            elif isinstance(types, list) and diagonal:
+                types = {'diagonal': types[0],
+                         'lower': types[0],
+                         'upper': types[-1]}
+            else:
+                types = {'lower': types[0], 'upper': types[-1]}
+        diagonal_kwargs = kwargs.pop('diagonal_kwargs', {})
+        lower_kwargs = kwargs.pop('lower_kwargs', {})
+        upper_kwargs = kwargs.pop('upper_kwargs', {})
+        diagonal_kwargs.update(kwargs)
+        lower_kwargs.update(kwargs)
+        upper_kwargs.update(kwargs)
```

Comment:
1. I believe that this implementation means that ``kwargs`` takes precedence over the ``diagonal_kwargs`` etc. I think it might be more intuitive if diagonal_kwargs overrode the kwargs entries
2. We could either have three separate arguments (as you have implemented here), or a dictionary of overrides (which would be more consistent with the types argument), taking ``'diagonal'``, ``'lower'`` and ``'upper'`` as keys. Either works, but we need to make a decision in this PR as to what the default should be. What do you reckon?

Reply:
1. This was actually an active decision of mine. I feel like anything directly passed to `plot_2d` should have precedence over something that gets indirectly passed to one subplot. Passing it directly kind of means "this is supposed to apply to all plots, not just one". For example if we passed `color` directly, it would ensure that the lines in the diagonal plots match in colour to the contours in the lower plots. If you really wanted different colours in the different plots, then you should specify it in all sub-kwargs and not pass it directly.

2. Making a dictionary here would turn it into a dictionary of dictionaries and I fear that would look rather ghastly...?

Reply:
1. OK. I agree with this reasoning. You should add a note in the documentation that kwargs overrides the individual ones.
2. Actual lol. OK, we should prioritise the user over implementation. You might find that you can clean up the code by constructing such a dictionary and then looping.

Reply:
1. done

2. Haha, three is not enough for a loop to justify the ugliness of dict in dict, here. Keeping it as is.

Reply:
OK. If we can get the tests passing and this merged, I'll submit a pull request that you can review that suggests how it might be made neater.